commit 234249d88b091d006b82f8d570343aae5f383736
Author: Wen Gong <quic_wgong@quicinc.com>
Date:   Fri Aug 25 03:00:55 2023 -0400

    wifi: cfg80211/mac80211: hold link BSSes when assoc fails for MLO connection
    
    When connect to MLO AP with more than one link, and the assoc response of
    AP is not success, then cfg80211_unhold_bss() is not called for all the
    links' cfg80211_bss except the primary link which means the link used by
    the latest successful association request. Thus the hold value of the
    cfg80211_bss is not reset to 0 after the assoc fail, and then the
    __cfg80211_unlink_bss() will not be called for the cfg80211_bss by
    __cfg80211_bss_expire().
    
    Then the AP always looks exist even the AP is shutdown or reconfigured
    to another type, then it will lead error while connecting it again.
    
    The detail info are as below.
    
    When connect with muti-links AP, cfg80211_hold_bss() is called by
    cfg80211_mlme_assoc() for each cfg80211_bss of all the links. When
    assoc response from AP is not success(such as status_code==1), the
    ieee80211_link_data of non-primary link(sdata->link[link_id]) is NULL
    because ieee80211_assoc_success()->ieee80211_vif_update_links() is
    not called for the links.
    
    Then struct cfg80211_rx_assoc_resp resp in cfg80211_rx_assoc_resp() and
    struct cfg80211_connect_resp_params cr in __cfg80211_connect_result()
    will only have the data of the primary link, and finally function
    cfg80211_connect_result_release_bsses() only call cfg80211_unhold_bss()
    for the primary link. Then cfg80211_bss of the other links will never free
    because its hold is always > 0 now.
    
    Hence assign value for the bss and status from assoc_data since it is
    valid for this case. Also assign value of addr from assoc_data when the
    link is NULL because the addrs of assoc_data and link both represent the
    local link addr and they are same value for success connection.
    
    Fixes: 81151ce462e5 ("wifi: mac80211: support MLO authentication/association with one link")
    Signed-off-by: Wen Gong <quic_wgong@quicinc.com>
    Link: https://lore.kernel.org/r/20230825070055.28164-1-quic_wgong@quicinc.com
    Signed-off-by: Johannes Berg <johannes.berg@intel.com>

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index f93eb38ae0b8..46d46cfab6c8 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -5429,17 +5429,18 @@ static void ieee80211_rx_mgmt_assoc_resp(struct ieee80211_sub_if_data *sdata,
 	for (link_id = 0; link_id < IEEE80211_MLD_MAX_NUM_LINKS; link_id++) {
 		struct ieee80211_link_data *link;
 
-		link = sdata_dereference(sdata->link[link_id], sdata);
-		if (!link)
-			continue;
-
 		if (!assoc_data->link[link_id].bss)
 			continue;
 
 		resp.links[link_id].bss = assoc_data->link[link_id].bss;
-		resp.links[link_id].addr = link->conf->addr;
+		ether_addr_copy(resp.links[link_id].addr,
+				assoc_data->link[link_id].addr);
 		resp.links[link_id].status = assoc_data->link[link_id].status;
 
+		link = sdata_dereference(sdata->link[link_id], sdata);
+		if (!link)
+			continue;
+
 		/* get uapsd queues configuration - same for all links */
 		resp.uapsd_queues = 0;
 		for (ac = 0; ac < IEEE80211_NUM_ACS; ac++)